### PR TITLE
Mutants Optimisations: big-packages, timeout multiplier, stacks-signer case

### PR DESCRIPTION
**Pull request by ASuciuX**
[Updated the workflow](https://github.com/stacks-network/actions/pull/23/files) to include the following changes: 
1. split big packages into `stackslib` and `stacks-node`
2. update cargo mutants to use specific version from github in order to include a timeout multiplier feature
3. add a specific run case for `stacks-signer` 
 
This PR needs to be merged beforehand: https://github.com/stacks-network/actions/pull/23.

The reason why I'm using `cargo nextest` before running the `stacks-signer` mutants is that on this specific package, the run has a bug: the test runs indefinitely, and then the workflow timeouts if `cargo nextest` is not run before the `cargo mutants` command.

Debug runs that showcase the issue that appeared when running `stacks-signer` mutants:
1. Run working with cargo build and cargo nextest run before cargo mutants:
Action run: https://github.com/ASuciuX/stacks-core/actions/runs/7901429286/attempts/2?pr=57
Action file: [https://github.com/stacks-network/actions/blob/8a7cb7ed0249b473d241b57228cd718d2bd1e3ef/stacks-core/mutation-testing/pr-differences/action.yml#L[…]9](https://github.com/stacks-network/actions/blob/8a7cb7ed0249b473d241b57228cd718d2bd1e3ef/stacks-core/mutation-testing/pr-differences/action.yml#L201-L209)
2. Run not working without cargo build and cargo nextest run:
Action run: https://github.com/ASuciuX/stacks-core/actions/runs/7901429286/attempts/3?pr=57
Action file: [https://github.com/stacks-network/actions/blob/801b21e8294c032686dab54b74cb579074141764/stacks-core/mutation-testing/pr-differences/action.yml#L[…]3](https://github.com/stacks-network/actions/blob/801b21e8294c032686dab54b74cb579074141764/stacks-core/mutation-testing/pr-differences/action.yml#L201-L203)
3. Run working with only cargo nextest run before cargo mutants:
Action run: https://github.com/ASuciuX/stacks-core/actions/runs/7901429286?pr=57
Action file: [https://github.com/stacks-network/actions/blob/b9971c8063623e9f4dee82164652264cce703216/stacks-core/mutation-testing/pr-differences/action.yml#L[…]6](https://github.com/stacks-network/actions/blob/b9971c8063623e9f4dee82164652264cce703216/stacks-core/mutation-testing/pr-differences/action.yml#L201-L206)

### CODE REVIEW

File: .github/workflows/pr-differences-mutants.yml [hunk 1]
```diff
@@ -68,15 +71,15 @@ jobs:
...
-  pr-differences-mutants-big-normal:
-    name: Mutation Testing - Normal, Big
+  # Mutation testing - Execute on PR on stackslib package (normal run, no shards)
+  pr-differences-mutants-stackslib-normal:
+    name: Mutation Testing - Normal, Stackslib
 
     needs: check-big-packages-and-shards
 
-    if: ${{ needs.check-big-packages-and-shards.outputs.run_big_packages == 'true' && needs.check-big-packages-and-shards.outputs.big_packages_with_shards == 'false' }}
+    if: ${{ needs.check-big-packages-and-shards.outputs.run_stackslib == 'true' && needs.check-big-packages-and-shards.outputs.stackslib_with_shards == 'false' }}
```
Comment (wileyj) [inline .github/workflows/pr-differences-mutants.yml:82]: can you change these conditionals so they're easier to read? ex:
https://github.com/stacks-network/stacks-core/blob/master/.github/workflows/ci.yml#L55-L63

this goes for any `if:` in this PR

**Comment by wileyj** [comment]
> 3\. https://github.com/ASuciuX/stacks-core/actions/runs/7901429286?pr=57

it looks like this build failed on `4: unmutated build failed` - should this be expected to pass?

**Comment by ASuciuX** [comment]
> > 3. https://github.com/ASuciuX/stacks-core/actions/runs/7901429286?pr=57
> 
> it looks like this build failed on `4: unmutated build failed` - should this be expected to pass?

Thanks for pointing that, it shouldn't be expected to fail... The `cargo nextest` passed it and then it failed on `cargo mutants running nextest`. I've messaged Marvin from `cargo mutants` regarding this because something is not working as expected.

File: .github/workflows/pr-differences-mutants.yml [hunk 2]
```diff
@@ -38,23 +41,27 @@ jobs:
 
     needs: check-big-packages-and-shards
 
-    if: ${{ needs.check-big-packages-and-shards.outputs.run_small_packages == 'true' && needs.check-big-packages-and-shards.outputs.small_packages_with_shards == 'false' }}
+    if: |
+      ${{ needs.check-big-packages-and-shards.outputs.run_small_packages == 'true' &&
+      needs.check-big-packages-and-shards.outputs.small_packages_with_shards == 'false' }}
```
Comment (wileyj) [inline .github/workflows/pr-differences-mutants.yml:46]: one of the benefits of moving to multiline conditionals is that you can remove the `${{ }}` chars, ex:
```suggestion
    if: |
      needs.check-big-packages-and-shards.outputs.run_small_packages == 'true' &&
      needs.check-big-packages-and-shards.outputs.small_packages_with_shards == 'false'
```

**Comment by ASuciuX** [comment]
This is the workflow that was run before. Now it is running as expected, without stacks-signer, even though there are changes on that package.

https://github.com/ASuciuX/stacks-core/actions/runs/8009445597/job/21878164232?pr=59

**Comment by wileyj** [comment]
> This is the workflow that was run before. Now it is running as expected, without stacks-signer, even though there are changes on that package.
> 
> https://github.com/ASuciuX/stacks-core/actions/runs/8009445597/job/21878164232?pr=59

and the output of "unviable mutants" is expected here?

**Comment by ASuciuX** [comment]
Yes, there are unviable mutants in this run. If it returns unviable, it means it finished the workflow as the problem was keeping the run till the whole github workflow timed out.

**Review by wileyj — APPROVED** [review]
shipit

File: .github/workflows/pr-differences-mutants.yml [hunk 3]
```diff
@@ -112,28 +123,100 @@ jobs:
...
+
+    needs: check-big-packages-and-shards
+
+    if: |
+      needs.check-big-packages-and-shards.outputs.run_stacks_signer == 'true'
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Run Run mutants on diffs
```
Comment (jbencin) [inline .github/workflows/pr-differences-mutants.yml:188]:
```suggestion
      - name: Run mutants on diffs
```

File: .github/workflows/pr-differences-mutants.yml [hunk 4]
```diff
@@ -112,28 +123,100 @@ jobs:
...
+        with:
+          package: 'stacks-signer'
 
   # Output the mutants and fail the workflow if there are missed/timeout/unviable mutants
   output-mutants:
     name: Output Mutants
 
     runs-on: ubuntu-latest
 
+    if: always()
```
Comment (jbencin) [inline .github/workflows/pr-differences-mutants.yml:199]: Unnecessary?

```suggestion
```

**Review by jbencin — DISMISSED** [review]

File: .github/workflows/pr-differences-mutants.yml (hunk 4 above)
Comment (ASuciuX) [reply .github/workflows/pr-differences-mutants.yml:199]: This is necessary to ensure that the output job continues to run even if a preceding job fails. This approach allows the output to clearly show the combined end result from the execution of the mutant jobs.

**Review by jbencin — APPROVED** [review]

**Review by wileyj — APPROVED** [review]

**Review by wileyj — APPROVED** [review]

**Review by wileyj — APPROVED** [review]

**Pull request merged**